crates/bevy_render/src/globals.rs

```diff
@@ -0,0 +1,70 @@
+use crate::{
+    extract_resource::ExtractResource,
+    render_resource::{ShaderType, UniformBuffer},
+    renderer::{RenderDevice, RenderQueue},
+    Extract, RenderApp, RenderStage,
+};
+use bevy_app::{App, Plugin};
+use bevy_ecs::prelude::*;
+use bevy_math::Vec4;
+use bevy_reflect::Reflect;
+use bevy_time::Time;
+
+pub struct GlobalsPlugin;
+
+impl Plugin for GlobalsPlugin {
+    fn build(&self, app: &mut App) {
+        if let Ok(render_app) = app.get_sub_app_mut(RenderApp) {
+            render_app
+                .init_resource::<GlobalsBuffer>()
+                .init_resource::<Time>()
+                .add_system_to_stage(RenderStage::Extract, extract_time)
+                .add_system_to_stage(RenderStage::Prepare, prepare_globals_buffer);
+        }
+    }
+}
+
+fn extract_time(mut commands: Commands, time: Extract<Res<Time>>) {
+    commands.insert_resource(time.clone());
+}
+
+#[derive(Default, Clone, ExtractResource, Reflect, ShaderType)]
+#[reflect(Resource)]
+pub struct GlobalsUniform {
+    time: Vec4,
+    dt: Vec4,
+}
+
+#[derive(Component, Default)]
+pub struct GlobalsBuffer {
+    pub buffer: UniformBuffer<GlobalsUniform>,
+}
+
+fn prepare_globals_buffer(
+    render_device: Res<RenderDevice>,
+    render_queue: Res<RenderQueue>,
+    mut globals_buffer: ResMut<GlobalsBuffer>,
+    time: Res<Time>,
+) {
+    let dt = time.delta_seconds();
+    let time_since_start = time.seconds_since_startup() as f32;
+
+    // This mimicks what Unity does
+    let time = Vec4::new(
+        time_since_start / 20.0,
+        time_since_start,
+        time_since_start * 2.0,
+        time_since_start * 3.0,
+    );
```

Comment:
I don’t remember what their justification was. Is it documented somewhere? We should document it.

Reply:
I didn't find anything about the justification. I guess it avoids a multiplication in shader code when you need them. I guess we could just start with time and dt as f32 for now and expand them when necessary.

Reply:
https://forums.developer.nvidia.com/t/float4-bandwidth-advantages-over-plain-float1/62799

Unity probably chose to use float4 for this reason (it's also a very old game engine, which ran on older architecture). A lot of the built in shader types also use float4 or other 128-bit sized types as well.

Reply:
The Vec4 makes sense, but the reason behind /20, *2 and *3 isn't clear.

Looking at other engines, godot, unreal and shadertoy all only provide a single f32 representing the time since start and godot resets the time to 0 after 1 hour. So I think I'll only keep a single f32 for now.

Reply:
Godot has a way to scale the time globally, but that feels a bit weird for it to be global.

Reply:
I spoke way too soon. shadertoy does provide a timeDelta too

Reply:
I guess we could pack time and dt in a single Vec4, but I'm not sure if it's worth it compared to the convenience of not needing to know what is in which part of the vec.